File: Algorithms/Implementation/ClimbingTheLeaderboard.py
#!/bin/python3
import math
import os
import random
import re
import sys

#
# Complete the 'climbingLeaderboard' function below.
#
# The function is expected to return an INTEGER_ARRAY.
# The function accepts following parameters:
#  1. INTEGER_ARRAY ranked
#  2. INTEGER_ARRAY player
#

def search(target, li, lenli):
    prefix = 0
    while True:
        if lenli <= 3:
            for i in range(lenli):
                if target >= li[i]:
                    return prefix + i
            return lenli + prefix
        isOdd = 0
        if lenli % 2 == 1:
            isOdd = 1
        #half = lenli // 2
        median = li[lenli // 2]
        if (target >= median and target < li[lenli // 2 - 1]):
            return lenli // 2 + prefix
        if target > median:
            li = li[:lenli // 2]
            lenli = lenli // 2
        else:
            prefix += lenli // 2 + 1
            li = li[lenli // 2 + 1:]
            lenli = lenli // 2 - 1 + isOdd
    """
    #lenli = len(li)
    if lenli <= 3:
        for i in range(lenli):
            if target >= li[i]:
                return i
        return lenli
    isOdd = 0
    if lenli % 2 == 1:
        isOdd = 1
    half = lenli // 2
    median = li[half]
    if (target >= median and target < li[half - 1]):
        return half
    if target > median:
        return search(target, li[:half], half)
    else:
        return half + search(target, li[half + 1:], half - 1 + isOdd) + 1
        """


def climbingLeaderboard(ranked, player):
    # Write your code here
    ranked = list(set(ranked))
    ranked.sort(reverse=True)
    retval = []
    lenran = len(ranked)
    lenpl = len(player)
    """
    for i in player:
        retval.append(1 + search(i, ranked, lenran))
        if retval[-1] - 1 == lenran or i != ranked[retval[-1] - 1]:
            ranked.insert(retval[-1] - 1, i)
            lenran += 1
    """
    player.reverse()
    curran = 0
    curpl = 0
    while curpl < lenpl and curran < lenran:
        if player[curpl] >= ranked[curran]:
            retval.insert(0, curran + 1)
            curpl += 1
        else:
            curran += 1
    for i in range(curpl, lenpl):
        retval.insert(0, lenran + 1)
    return retval

if __name__ == '__main__':
    #fptr = open(os.environ['OUTPUT_PATH'], 'w')

    ranked_count = int(input().strip())

    ranked = list(map(int, input().rstrip().split()))

    player_count = int(input().strip())

    player = list(map(int, input().rstrip().split()))

    result = climbingLeaderboard(ranked, player)
    print('\n'.join(map(str, result)))
    print('\n')

    #fptr.write('\n'.join(map(str, result)))
    #fptr.write('\n')

    #fptr.close()

File: Algorithms/Implementation/ClimbingTheLeaderboard.cpp
#include <vector>
#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>

using namespace std;

string ltrim(const string &);
string rtrim(const string &);
vector<string> split(const string &);

/*
 * Complete the 'climbingLeaderboard' function below.
 *
 * The function is expected to return an INTEGER_ARRAY.
 * The function accepts following parameters:
 *  1. INTEGER_ARRAY ranked
 *  2. INTEGER_ARRAY player
 */

int bs(const vector<int>& arr, int n, int bsR)
{
    int bsLeft = 0, bsRight = bsR, bsCurrent = (bsLeft + bsRight) / 2;
    while (!(n >= arr[bsCurrent] && n < arr[bsCurrent - 1]))
    {
        if (bsLeft == bsRight)
            return bsLeft;
        if (bsLeft > bsRight)
            return -1;
        if (bsRight - bsLeft == 1)
        {
            if (n >= arr[bsRight] && n < arr[bsLeft])
                return bsRight;
            else if (n >= arr[bsLeft] && n < arr[bsLeft - 1])
                return bsLeft;
            else
                return -1;
        }
        if (n == arr[bsCurrent])
            return bsCurrent;
        if (n >= arr[bsCurrent])
            bsRight = bsCurrent;
        else
            bsLeft = bsCurrent;
        bsCurrent = (bsLeft + bsRight) / 2;
    }
    return bsCurrent;
}
int bsRecursive(const vector<int>& arr, const int& n)
{
    if (arr.size() <= 2)
    {
        if (arr[0] <= n)
            return 0;
        else
            return 1;
    }
    int current = arr.size() / 2 - !(arr.size() % 2);
    if (arr[current] >= n && n < arr[current - 1])
        return current;
    if (n >= arr[current])
        return bsRecursive(vector<int>(arr.cbegin(), (arr.size() % 2) + arr.cend() - current), n);
    else
        return 1 + current + bsRecursive(vector<int>(arr.cbegin() + current + (arr.size() % 2) + 1, arr.cend()), n);
}

vector<int> climbingLeaderboard(vector<int> ranked, vector<int> player) {
    vector<int> retval;
    sort(ranked.begin(), ranked.end());
    ranked.erase(unique(ranked.begin(), ranked.end()), ranked.end());
    reverse(ranked.begin(), ranked.end());
    //int prevIndex = ranked.size() - 1;
    for (auto i : player)
    {
        if (i <= ranked.back())
        {
            if (i != ranked.back())
                ranked.push_back(i);
            retval.push_back(ranked.size());
            //prevIndex = ranked.size() - 1;
            continue;
        }
        if (i >= ranked[0])
        {
            if (i != ranked[0])
            {
                ranked.clear();
                ranked.push_back(i);
            }
            retval.push_back(1);
            //prevIndex = 0;
            continue;
        }
        auto bsCurrent = /*bs(ranked, i, prevIndex)*/bsRecursive(ranked, i);
        ranked.resize(bsCurrent + 1);
        if (i != ranked.back())
        {
            ranked.pop_back();
            ranked.push_back(i);
        }
        retval.push_back(bsCurrent + 1);
        //prevIndex = bsCurrent;
    }
    return retval;
}


int main()
{
    ofstream fout(getenv("OUTPUT_PATH"));

    string ranked_count_temp;
    getline(cin, ranked_count_temp);

    int ranked_count = stoi(ltrim(rtrim(ranked_count_temp)));

    string ranked_temp_temp;
    getline(cin, ranked_temp_temp);

    vector<string> ranked_temp = split(rtrim(ranked_temp_temp));

    vector<int> ranked(ranked_count);

    for (int i = 0; i < ranked_count; i++) {
        int ranked_item = stoi(ranked_temp[i]);

        ranked[i] = ranked_item;
    }

    string player_count_temp;
    getline(cin, player_count_temp);

    int player_count = stoi(ltrim(rtrim(player_count_temp)));

    string player_temp_temp;
    getline(cin, player_temp_temp);

    vector<string> player_temp = split(rtrim(player_temp_temp));

    vector<int> player(player_count);

    for (int i = 0; i < player_count; i++) {
        int player_item = stoi(player_temp[i]);

        player[i] = player_item;
    }

    vector<int> result = climbingLeaderboard(ranked, player);

    for (int i = 0; i < result.size(); i++) {
        fout << result[i];

        if (i != result.size() - 1) {
            fout << "\n";
        }
    }

    fout << "\n";

    fout.close();

    return 0;
}

string ltrim(const string &str) {
    string s(str);

    s.erase(
        s.begin(),
        find_if(s.begin(), s.end(), not1(ptr_fun<int, int>(isspace)))
    );

    return s;
}

string rtrim(const string &str) {
    string s(str);

    s.erase(
        find_if(s.rbegin(), s.rend(), not1(ptr_fun<int, int>(isspace))).base(),
        s.end()
    );

    return s;
}

vector<string> split(const string &str) {
    vector<string> tokens;

    string::size_type start = 0;
    string::size_type end = 0;

    while ((end = str.find(" ", start)) != string::npos) {
        tokens.push_back(str.substr(start, end - start));

        start = end + 1;
    }

    tokens.push_back(str.substr(start));

    return tokens;
}

